File: python38/src/geometry/coefficient_base.py
import numpy as np
from abstract import Abstract

class CoefficientBase(Abstract):

    def __init__(self, order):
        self._order = order

    @property
    def order(self) -> int:
        return self._order

    def t_coefficient(self, t):
        res = np.zeros(self._order + 1)
        res[0] = 1
        for i in range(1, res.shape[0]):
            res[i] = res[i-1] * t
        return res

    def t_first_derivative_coefficient(self, t):
        res = np.zeros(self._order + 1)
        t_coeff = self.t_coefficient(t)
        for i in range(1, res.shape[0]):
            res[i] = t_coeff[i-1] * i
        return res

    def t_second_derivative_coefficient(self, t):
        res = np.zeros(self._order + 1)
        t_first_derivative_coeff = self.t_first_derivative_coefficient(
            t)
        for i in range(2, res.shape[0]):
            res[i] = t_first_derivative_coeff[i-1] * i
        return res

    def t_third_derivative_coefficient(self, t):
        res = np.zeros(self._order + 1)
        t_second_derivative_coeff = self.t_second_derivative_coefficient(
            t)
        for i in range(3, res.shape[0]):
            res[i] = t_second_derivative_coeff[i-1] * i
        return res

File: cpp20/src/geometry/coefficient_base.cc
#include "coefficient_base.h"

using namespace Geometry;


std::vector<double> CoefficientBase::t_coefficient(double t) const {
  std::vector<double> res(_order + 1, 0);
  res[0] = 1;
  for (int i = 1; i <= _order; ++i) {
    res[i] = res[i - 1] * t;
  }
  return res;
}

std::vector<double>
CoefficientBase::t_first_derivative_coefficient(double t) const {
  std::vector<double> res(_order + 1, 0);
  auto t_coeff = t_coefficient(t);
  for (int i = 1; i <= _order; ++i) {
    res[i] = t_coeff[i - 1] * i;
  }
  return res;
}

std::vector<double>
CoefficientBase::t_second_derivative_coefficient(double t) const {
  std::vector<double> res(_order + 1, 0);
  auto t_first_derivative_coeff = t_first_derivative_coefficient(t);
  for (int i = 2; i <= _order; ++i) {
    res[i] = t_first_derivative_coeff[i - 1] * i;
  }
  return res;
}

std::vector<double>
CoefficientBase::t_third_derivative_coefficient(double t) const {
  std::vector<double> res(_order + 1, 0);
  auto t_second_derivative_coeff = t_second_derivative_coefficient(t);
  for (int i = 3; i <= _order; ++i) {
    res[i] = t_second_derivative_coeff[i - 1] * i;
  }
  return res;
}